06_HelloGraphicsQueue/main.cpp

```diff
@@ -2,80 +2,436 @@
+
+		IGPUCommandBuffer::SImageMemoryBarrier smallImgSecondBlitBarrier = constructDefaultInitializedImageBarrier();
+		smallImgSecondBlitBarrier.barrier.srcAccessMask = E_ACCESS_FLAGS::EAF_MEMORY_WRITE_BIT;
+		smallImgSecondBlitBarrier.barrier.dstAccessMask = E_ACCESS_FLAGS::EAF_MEMORY_READ_BIT;
+		smallImgSecondBlitBarrier.oldLayout = asset::IImage::EL_TRANSFER_DST_OPTIMAL;
+		smallImgSecondBlitBarrier.newLayout = asset::IImage::EL_TRANSFER_SRC_OPTIMAL;
+		smallImgSecondBlitBarrier.image = smallImg;
+
+		cmdbuf->pipelineBarrier(
+			E_PIPELINE_STAGE_FLAGS::EPSF_TRANSFER_BIT,
+			E_PIPELINE_STAGE_FLAGS::EPSF_TRANSFER_BIT,
+			E_DEPENDENCY_FLAGS::EDF_NONE, 0u, nullptr, 0u, nullptr, 1u, &smallImgSecondBlitBarrier
+		);
+
+		// Copy the resulting image to a buffer.
+		cmdbuf->copyImageToBuffer(smallImg.get(), IImage::E_LAYOUT::EL_TRANSFER_SRC_OPTIMAL, outputImageBuffer.get(), 1u, &bufferCopy);
+
+		cmdbuf->end();
+
+		queue->startCapture();
+		queue->submit(1u, &submitInfo, done.get());
+		queue->endCapture();
+		m_device->blockForFences(1u, &done.get());
+
+		// Read the buffer back, create an ICPUImage with an adopted buffer over its contents.
+
+		// Map memory, so contents of `outputImageBuffer` will be host visible.
+		const IDeviceMemoryAllocation::MappedMemoryRange memoryRange(outputBufferAllocation.memory.get(), 0ull, outputBufferAllocation.memory->getAllocationSize());
+		auto imageBufferMemPtr = m_device->mapMemory(memoryRange, IDeviceMemoryAllocation::EMCAF_READ);
+		if (!imageBufferMemPtr)
+			return logFail("Failed to map the Device Memory!\n");
+
+		// If the mapping is not coherent the range needs to be invalidated to pull in new data for the CPU's caches.
+		if (!outputBufferAllocation.memory->getMemoryPropertyFlags().hasFlags(IDeviceMemoryAllocation::EMPF_HOST_COHERENT_BIT))
+			m_device->invalidateMappedMemoryRanges(1, &memoryRange);
+
+		const auto ouputImageCreationParams = smallImg->getCreationParameters();
+		// While JPG/PNG/BMP/EXR Loaders create ICPUImages because they cannot disambiguate colorspaces,
+		// 2D_ARRAY vs 2D and even sometimes formats (looking at your PNG normalmaps!),
+		// the writers are always meant to be fed by ICPUImageViews.
+		ICPUImageView::SCreationParams params = {};
+		{
+			params.image = ICPUImage::create(ouputImageCreationParams);
+			{
+				auto cpuView = core::make_smart_refctd_ptr<CDummyCPUBuffer>(smallImgByteSize, imageBufferMemPtr, core::adopt_memory_t());
+				ICPUImage::SBufferCopy region = {};
+				region.imageSubresource.aspectMask = IImage::E_ASPECT_FLAGS::EAF_COLOR_BIT;
+				region.imageSubresource.layerCount = 1;
+				region.imageExtent = ouputImageCreationParams.extent;
+				params.image->setBufferAndRegions(std::move(cpuView),core::make_refctd_dynamic_array<core::smart_refctd_dynamic_array<ICPUImage::SBufferCopy>>(1,region));
+			}
```

Comment:
this needs a lot of comments to explain whats going on

Reply:
basically explain that our ICPUImage isn't really a representation of a GPU Image in itself, more of a recipe for creating one from a series of ICPUBuffer to ICPUImage copies.

This means that an ICPUImage has no internal storage or memory bound for its texels and rather references separate ICPUBuffer ranges to provide its contents, which also means it can be sparsely (with gaps) specified.